### PR TITLE
feat: add initial concepts and replace sfinae

### PR DESCRIPTION
**Pull request by Shivix**
Adds some basic concepts and uses them to replace instances of enable_if
style sfinae.

SDB-3166

### CODE REVIEW

File: toolbox/util/String.hpp [hunk 1]
```diff
@@ -51,14 +51,20 @@ constexpr ValueT from_string(const char* s)
...
 
-template <typename ValueT, typename std::enable_if_t<!std::is_arithmetic_v<ValueT>>* = nullptr>
+template <typename ValueT>
+// clang-format off
+requires (!Arithmetic<ValueT>)
 std::string to_string(const ValueT& val)
+// clang-format on
 {
     std::stringstream ss;
     ss << val;
```
Comment (Shivix) [inline toolbox/util/String.hpp:70]: I could potentially rewrite this with the following:
```
template <typename ValueT>
std::string to_string(ValueT&& val)
{
    std::stringstream ss;
    ss << val;
    return ss.str();
}

template <typename ValueT>
// clang-format off
requires Arithmetic<ValueT>
std::string to_string(ValueT&& val)
// clang-format on
{
    return std::to_string(val);
}
```
But we'll lose copy semantics on the Arithmetic `to_string`
Not worth it?
A [CppCoreGuideline](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-not) on the issue

File: toolbox/net/Endian.hpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #ifndef TOOLBOX_NET_ENDIAN_HPP
 #define TOOLBOX_NET_ENDIAN_HPP
 
+#include <concepts>
```
Comment (markaylett) [inline toolbox/net/Endian.hpp:20]: Conventionally, this should go below `<bit>`.

File: toolbox/net/Endian.hpp [hunk 3]
```diff
@@ -57,7 +58,8 @@ constexpr std::int64_t bswap(std::int64_t n) noexcept
     return bswap(static_cast<std::uint64_t>(n));
 }
 
-template <typename ValueT, typename = std::enable_if_t<std::is_integral_v<ValueT>>>
+template <typename ValueT>
+requires std::integral<ValueT>
```
[the remark this answers is not in the document]
Comment (markaylett) [reply toolbox/net/Endian.hpp:62]: I personally have no strong preference either way.

File: toolbox/util/String.hpp (hunk 1 above)
Comment (markaylett) [reply toolbox/util/String.hpp:70]: Let's discuss this one on a call.

**Review by markaylett — APPROVED** [review]

File: toolbox/net/Endian.hpp (hunk 2 above)
Comment (Shivix) [reply toolbox/net/Endian.hpp:20]: Ah good spot, I actually would have thought clang-format would have fixed this to be honest.

**Pull request merged**